app/src/behaviors/behavior_macro.c

```diff
@@ -0,0 +1,192 @@
+/*
+ * Copyright (c) 2022 The ZMK Contributors
+ *
+ * SPDX-License-Identifier: MIT
+ */
+
+#define DT_DRV_COMPAT zmk_behavior_macro
+
+#include <device.h>
+#include <drivers/behavior.h>
+#include <logging/log.h>
+#include <zmk/behavior.h>
+#include <zmk/behavior_queue.h>
+#include <zmk/keymap.h>
+
+LOG_MODULE_DECLARE(zmk, CONFIG_ZMK_LOG_LEVEL);
+
+#if DT_HAS_COMPAT_STATUS_OKAY(DT_DRV_COMPAT)
+
+enum behavior_macro_mode {
+    MACRO_MODE_TAP,
+    MACRO_MODE_PRESS,
+    MACRO_MODE_RELEASE,
+};
+
+struct behavior_macro_trigger_state {
+    uint32_t wait_ms;
+    uint32_t tap_ms;
+    enum behavior_macro_mode mode;
+    int8_t start_index;
+    int8_t end_index;
+};
+
+struct behavior_macro_state {
+    struct behavior_macro_trigger_state release_state;
+
+    int8_t last_press_binding;
+};
+
+struct behavior_macro_config {
+    uint32_t default_wait_ms;
+    uint32_t default_tap_ms;
+    uint8_t count;
```

Comment:
Using a uint8 over a uint32 probably doesn't save any memory space unless you explicitly pack this struct. The compiler is most likely going to pad this out to 32-bits.

Reply:
Tweaked a bit.